test/hotspot/jtreg/runtime/Monitor/MonitorWithDeadObjectTest.java

```diff
@@ -0,0 +1,101 @@
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+/*
+ * @bug 8320515
+ * @summary This test checks that ObjectMonitors with dead objects don't
+ *          cause asserts, crashes, or failures when various sub-systems
+ *          in the JVM find them.
+ * @library /testlibrary /test/lib
+ * @modules jdk.management
+ */
+
+/*
+ * @requires os.family != "windows" & os.family != "aix"
+ * @test id=DetachThread
+ * @run main/othervm/native MonitorWithDeadObjectTest 0
+ */
+
+/*
+ * @requires os.family != "windows" & os.family != "aix"
+ * @test id=DumpThreadsBeforeDetach
+ * @run main/othervm/native MonitorWithDeadObjectTest 1
+ */
+
+/*
+ * @requires os.family != "windows" & os.family != "aix"
+ * @test id=DumpThreadsAfterDetach
+ * @run main/othervm/native MonitorWithDeadObjectTest 2
+ */
+
+import java.lang.management.ManagementFactory;
+import java.lang.management.ThreadMXBean;
+
+public class MonitorWithDeadObjectTest {
+    public static native void createMonitorWithDeadObject();
+    public static native void createMonitorWithDeadObjectDumpThreadsBeforeDetach();
+
+    static {
+        System.loadLibrary("MonitorWithDeadObjectTest");
+    }
+
+    private static void dumpThreadsWithLockedMonitors() {
+        ThreadMXBean threadBean = ManagementFactory.getThreadMXBean();
+        threadBean.dumpAllThreads(true, false);
+    }
+
+    private static void testDetachThread() {
+        // Create an ObjectMonitor with a dead object from an
+        // attached thread. This used to provoke an assert
+        // in DetachCurrentThread.
+        createMonitorWithDeadObject();
+    }
+
+    private static void testDumpThreadsBeforeDetach() {
+        // Create an ObjectMonitor with a dead object from an
+        // attached thread and perform a thread dump before
+        // detaching the thread.
+        createMonitorWithDeadObjectDumpThreadsBeforeDetach();
+    }
+
+    private static void testDumpThreadsAfterDetach() {
+        createMonitorWithDeadObject();
+
+        // After createMonitorWithDeadObject has been called, there's an "owned"
+        // monitor with a dead object. The thread dumping code used to not
+        // tolerate such a monitor and would assert. Run a thread dump and make
+        // sure that it doesn't crash/assert.
```

Comment:
The comment is not correct, the monitor will be unlocked by the time `createMonitorWithDeadObject` has returned. This post-dump is really just a sanity test.

Reply:
All of these are sanity checks now that this PR fixes the bug.

This specific test was added because of the combinations of bugs I've seen and provoked by temporarily reinstating various combinations of the bugs. Specifically, if the detach code skips visiting monitors with dead objects, but the thread dumping code does. That is, the opposite of the currently proposed patch.

I'll update the comment.